pkg/odo/cli/service/create.go

```diff
@@ -447,56 +477,50 @@ func NewCmdServiceCreate(name, fullName string) *cobra.Command {
 	return serviceCreateCmd
 }
 
-// Parses group and version values from the alm-example
-func groupVersionALMExample(example map[string]interface{}) (group, version string) {
-	apiVersion := example["apiVersion"].(string)
-	// use SplitN so that if apiVersion field's value is something like
-	// etcd.coreos.com/v1/beta1 then group's value ends up being etcd.cores.com
-	// and version ends up being v1/beta1
-	gv := strings.SplitN(apiVersion, "/", 2)
-
-	group, version = gv[0], gv[1]
-	return
-}
-
-func resourceFromCSV(csv olmv1alpha1.ClusterServiceVersion, crdName string) (resource string) {
-	for _, crd := range csv.Spec.CustomResourceDefinitions.Owned {
-		if crd.Kind == crdName {
-			resource = strings.Split(crd.Name, ".")[0]
-			return
-		}
+// validateMetadataInCRD validates if the CRD has metadata.name field and returns an error
+func (d *DynamicCRD) validateMetadataInCRD() error {
+	metadata, ok := d.OriginalCRD["metadata"].(map[string]interface{})
+	if !ok {
+		// this condition is satisfied if there's no metadata at all in the provided CRD
+		return fmt.Errorf("Couldn't find \"metadata\" in the yaml. Need metadata.name to start the service")
 	}
-	return
-}
 
-func getAlmExample(almExamples []map[string]interface{}, crd, operator string) (map[string]interface{}, error) {
-	for _, example := range almExamples {
-		if example["kind"].(string) == crd {
-			return example, nil
-		}
+	if _, ok := metadata["name"].(string); ok {
+		// found the metadata.name; no error
+		return nil
 	}
-	return nil, errors.Errorf("Could not find example yaml definition for %q service in %q operator's definition.\nPlease provide a file containing yaml specification to start the service from operator\n", crd, operator)
+	return fmt.Errorf("Couldn't find metadata.name in the yaml. Provide a name for the service")
 }
 
-func doesCRExist(kind string, csvs *olmv1alpha1.ClusterServiceVersionList) (olmv1alpha1.ClusterServiceVersion, error) {
-	for _, csv := range csvs.Items {
-		for _, operatorCR := range csv.Spec.CustomResourceDefinitions.Owned {
-			if kind == operatorCR.Kind {
-				return csv, nil
-			}
+// setServiceName modifies the CRD to contain user provided name on the CLI
+// instead of using the default one in almExample
+func (d *DynamicCRD) setServiceName(name string) error {
+	m := d.OriginalCRD["metadata"].(map[string]interface{})
+
+	for k := range m {
+		if k == "name" {
+			m[k] = name
+			return nil
 		}
+		// if metadata doesn't have 'name' field, we set it up
+		m["name"] = name
 	}
-	return olmv1alpha1.ClusterServiceVersion{}, errors.New("Could not find the requested cluster resource")
 
+	// if we reach this point, it's likely becuase the CRD doesn't have
+	// metadata or doesn't have metadata.name
```

Comment:
But if it doesn't have `metadata.name`, we are setting it up above, right? Or am I missing something?

Reply:
True. I feel it is necessary though as it catches the case where `metadata` is missing. If its there and name is not set, it is set here so comment needs fixing

Reply:
@mohammedzee1000 +1.

Removed this part of the code since `metadata` is surely going to be there in an alm example. If it's the case of doing `--from-file`, we're now validating the yaml before setting name. That should suffice, IMO.

If you folks think I'm missing a case, let me know.